src/auth/index.tsx

```diff
@@ -0,0 +1,142 @@
+import React, { useEffect, useState } from "react";
+import {
+  IAuthState,
+  IAuthProviderProps,
+  IAdvertiser,
+} from "auth/context/auth.interface";
+import { initialState, IAuthContext } from "auth/context/auth.state";
+import { getAdvertisers } from "./advertiser";
+import _ from "lodash";
+import { clearCredentials, getUser } from "./util";
+import { UserFragment } from "../graphql/user.generated";
+
+export const IAuthProvider: React.FC<IAuthProviderProps> = ({
+  children,
+}: IAuthProviderProps) => {
+  const [state, setState] = useState<IAuthState>(initialState);
+  const [loading, setLoading] = useState(false);
+
+  const setActiveAdvertiser = (advertiser?: IAdvertiser) => {
+    if (advertiser) {
+      window.localStorage.setItem(
+        "activeAdvertiser",
+        JSON.stringify(advertiser)
+      );
+
+      setState((cur) => ({
+        ...cur,
+        advertiser,
+      }));
+    }
+  };
+
+  const setSessionUser = (u?: UserFragment) => {
+    if (u) {
+      setLoading(true);
+      setState((cur) => ({
+        ...cur,
+        email: u.email,
+        emailVerified: u.emailVerified,
+        role: u.role,
+        userId: u.id,
+      }));
+
+      advertiser(u.id).finally(() => {
+        setLoading(false);
+      });
+    } else {
+      setState((cur) => ({
+        ...cur,
+        isAuthenticated: false,
+      }));
+    }
+  };
+
+  const getActiveAdvertiser = (): IAdvertiser | null => {
+    const adv = window.localStorage.getItem("activeAdvertiser");
+
+    if (adv) {
+      const parsed = JSON.parse(adv);
```

Comment:
Can we just store the advertiser id in local storage, rather than the whole object?

This gives the impression that I can e.g. enable editing by hacking the object in local storage. I don't think that works, as the code below seems to overwrite the object, and in any case we enforce server-side so it's not a security risk. 

But to avoid the next person having to do that same investigation, let's not store the whole object if possible.

Reply:
>  I don't think that works, as the code below seems to overwrite the object

Actually it doesn't, so this is a way to enable the UI even though the server will reject those changes. Let's not rely on untrusted data in localStorage like this.